test/core/security/tls_security_connector_test.cc

```diff
@@ -737,6 +737,44 @@ TEST_F(TlsSecurityConnectorTest,
                             on_peer_checked);
 }
 
+TEST_F(TlsSecurityConnectorTest,
+       ChannelSecurityConnectorWithVerifiedRootCertSubjectSucceeds) {
+  auto* sync_verifier_ = new SyncExternalVerifier(true);
+  ExternalCertificateVerifier core_external_verifier(sync_verifier_->base());
+  RefCountedPtr<grpc_tls_credentials_options> options =
+      MakeRefCounted<grpc_tls_credentials_options>();
+  options->set_verify_server_cert(true);
+  options->set_certificate_verifier(core_external_verifier.Ref());
+  options->set_check_call_host(false);
+  RefCountedPtr<TlsCredentials> credential =
+      MakeRefCounted<TlsCredentials>(options);
+  ChannelArgs new_args;
+  RefCountedPtr<grpc_channel_security_connector> connector =
+      credential->create_security_connector(nullptr, kTargetName, &new_args);
+  EXPECT_NE(connector, nullptr);
+  TlsChannelSecurityConnector* tls_connector =
+      static_cast<TlsChannelSecurityConnector*>(connector.get());
+  EXPECT_NE(tls_connector->ClientHandshakerFactoryForTesting(), nullptr);
+  // Construct a basic TSI Peer.
+  char* expected_subject = const_cast<char*>(
+      "CN=testca,O=Internet Widgits Pty Ltd,ST=Some-State,C=AU");
+  tsi_peer peer;
+  GPR_ASSERT(tsi_construct_peer(2, &peer) == TSI_OK);
+  GPR_ASSERT(tsi_construct_string_peer_property(TSI_SSL_ALPN_SELECTED_PROTOCOL,
+                                                "grpc", strlen("grpc"),
+                                                &peer.properties[0]) == TSI_OK);
+  GPR_ASSERT(tsi_construct_string_peer_property_from_cstring(
+                 TSI_X509_VERIFIED_ROOT_CERT_SUBECT_PEER_PROPERTY,
+                 expected_subject, &peer.properties[1]) == TSI_OK);
+  RefCountedPtr<grpc_auth_context> auth_context;
+  ExecCtx exec_ctx;
+  grpc_closure* on_peer_checked = GRPC_CLOSURE_CREATE(
+      VerifyExpectedErrorCallback, nullptr, grpc_schedule_on_exec_ctx);
+  ChannelArgs args;
+  tls_connector->check_peer(peer, nullptr, args, &auth_context,
+                            on_peer_checked);
+}
```

Comment:
IIUC, we're trying to test the `PendingVerifierRequestInit` API. Where/how is that called by the above test?

Reply:
`PendingVerifierRequestInit` is only in the cpp file, not the header.

It's called indirectly through `connector->check_peer`. 

The stack is short - it's called [during the constructor of {Channel/Server}PendingVerifierRequest}](https://github.com/grpc/grpc/blob/d5fcbce4b4ff81f246bf63899f51f5a3bd56fb8c/src/core/lib/security/security_connector/tls/tls_security_connector.cc#L476), and those are created [during the associated check_peer](https://github.com/grpc/grpc/blob/d5fcbce4b4ff81f246bf63899f51f5a3bd56fb8c/src/core/lib/security/security_connector/tls/tls_security_connector.cc#L370).

I could simplify the test and just make sure that we can construct the `{Channel/Server}PendingVerifierRequest` object since `PendingVerifierRequestInit` is called during their constructors. Thoughts?

Reply:
This might be fine, I just want to make sure I understand. When were calling the `check_peer` API in this test, how do we know whether or not the peer property was present when we expect it to be present?

Reply:
It's a good point. Right now the test essentially just makes sure it doesn't crash when the property is present, but it doesn't do anything to verify the actual value.

The issue is, it's all inside of private variables - the `{Channel/Server}PendingVerifierRequest` are private, and all of this happens within `check_peer`. The request is created, placed in a map `pending_verifier_requests_`, then removed from that map when completed, all without external accessibility.

I'll keep thinking on how to improve this, but I am open to suggestions.

Reply:
Could we leverage the VerifiedRootCertSubjectVerifier to do this?

(Apologies again if I'm missing something obvious.)

Reply:
That's a test util added in `test/cpp` and uses the cpp API. This test we're looking at is a level lower in `test/core` in the `security_connector`

Unfortunately that higher level cpp test doesn't cover either - we want to make sure that, given a `tsi_peer`, the `security_connector` level code can construct a `grpc_tls_custom_verification_check_request` that properly pulls the information from the `tsi_peer`.

The cpp level test that uses the `VerifiedRootCertSubjectVerifier` starts by constructing a `grpc_tls_custom_verification_check_request` itself, so we still doesn't have coverage of the `tsi_peer` -> `grpc_tls_custom_verification_check_request`.

I could create a public test-only wrapper function for `PendingVerifierRequestInit`, and I think that would help us test this? Thoughts?

Reply:
Would it be more convenient to test this in an end2end test?

Reply:
+1, good call Luwei. Let's modify the e2e tests for custom verification to check this.

Reply:
Added an E2E test, PTAL

I named the custom verifier `TestExternalVerifier` instead of something specific to root cert subject. I did this so that we have a modifiable custom verifier for checking any property if it is something we want to expand in the future.